File: part03/scala/src/main/scala/com/waitingforcode/Exercise2DistinctVisitsJob.scala
package com.waitingforcode

import org.apache.spark.sql.SparkSession

object Exercise2DistinctVisitsJob {

  def main(args: Array[String]): Unit = {
    val sparkSession = SparkSession.builder().master("local[*]")
      .getOrCreate()

    val inputDataStream = sparkSession.readStream
      .options(Map(
        "kafka.bootstrap.servers" -> "localhost:9094",
        "subscribe" -> InputDataTopicName,
        "startingOffsets" -> "EARLIEST",
      ))
      .format("kafka").load()

    import sparkSession.implicits._
    val query = inputDataStream.selectExpr("CAST(value AS STRING)")

    val writer = query.writeStream.option("checkpointLocation", "/tmp/wfc/workshop/part03/checkpoint/deduplication")
      .format("kafka")
      .option("kafka.bootstrap.servers", "localhost:9094")
      .option("topic", DedupedDataTopicName)

    writer.start().awaitTermination()
  }

}

File: part03/python/exercise2_distinct_visits_job.py
from pyspark.sql import SparkSession, functions
from pyspark.sql.types import StructType, StructField, StringType, TimestampType

from config import get_checkpoint_location, kafka_input_topic, kafka_deduped_topic

spark = SparkSession.builder.master("local[*]") \
    .config('spark.jars.packages', 'org.apache.spark:spark-sql-kafka-0-10_2.12:3.5.0') \
    .getOrCreate()

input_data = spark.readStream \
    .format("kafka") \
    .option("kafka.bootstrap.servers", "localhost:9094") \
    .option("subscribe", kafka_input_topic()) \
    .option("startingOffsets", "EARLIEST") \
    .load()

query = input_data.selectExpr("CAST(value AS STRING)")

writer = query.writeStream.option("checkpointLocation", get_checkpoint_location('deduplication'))\
    .format("kafka")\
    .option("kafka.bootstrap.servers", "localhost:9094")\
    .option("topic", kafka_deduped_topic())

writer.start().awaitTermination()
